// @react-mapboxgl/core/lib/LayerEvents.js
'use strict';

exports.__esModule = true;

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _omit2 = require('lodash/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _keys2 = require('lodash/keys');

var _keys3 = _interopRequireDefault(_keys2);

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Children = require('./Children');

var _Children2 = _interopRequireDefault(_Children);

var _LayerEvent = require('./LayerEvent');

var _LayerEvent2 = _interopRequireDefault(_LayerEvent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LayerEvents = (_temp = _class = function (_React$PureComponent) {
  _inherits(LayerEvents, _React$PureComponent);

  function LayerEvents() {
    _classCallCheck(this, LayerEvents);

    return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
  }

  LayerEvents.pickEvents = function pickEvents(props) {
    return (0, _pick3.default)(props, (0, _keys3.default)((0, _omit3.default)(LayerEvents.propTypes, 'layer')));
  };

  LayerEvents.prototype.render = function render() {
    var _this2 = this;

    return _react2.default.createElement(
      _Children2.default,
      null,
      (0, _map3.default)(LayerEvents.propTypes, function (_, type) {
        return _this2.props[type] && type !== 'layer' ? _react2.default.createElement(_LayerEvent2.default, {
          key: type,
          type: type.slice(2).toLowerCase(),
          layer: _this2.props.layer,
          onChange: _this2.props[type]
        }) : null;
      })
    );
  };

  return LayerEvents;
}(_react2.default.PureComponent), _class.propTypes = {
  // Layer id.
  layer: _propTypes2.default.string.isRequired,

  // Events.
  onDblClick: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  onMouseMove: _propTypes2.default.func,
  onMouseEnter: _propTypes2.default.func,
  onMouseLeave: _propTypes2.default.func,
  onMouseOver: _propTypes2.default.func,
  onMouseOut: _propTypes2.default.func,
  onMouseDown: _propTypes2.default.func,
  onMouseUp: _propTypes2.default.func,
  onTouchStart: _propTypes2.default.func,
  onTouchEnd: _propTypes2.default.func,
  onTouchCancel: _propTypes2.default.func,
  onContextMenu: _propTypes2.default.func
}, _temp);
exports.default = LayerEvents;
